environment/instance.go

```diff
@@ -0,0 +1,336 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package environment
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3em/build"
+	"github.com/m3db/m3em/operator"
+
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/shard"
+	m3dbrpc "github.com/m3db/m3db/generated/thrift/rpc"
+	m3dbchannel "github.com/m3db/m3db/network/server/tchannelthrift/node/channel"
+	tchannel "github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/thrift"
+)
+
+var (
+	errUnableToSetupInitializedInstance = fmt.Errorf("unable to setup instance, must be either setup/uninitialized")
+	errUnableToResetInstance            = fmt.Errorf("unable to reset instance, must be either setup/running")
+	errUnableToTeardownInstance         = fmt.Errorf("unable to teardown instance, must be either setup/running")
+	errUnableToStartInstance            = fmt.Errorf("unable to start instance, it must be setup")
+	errUnableToStopInstance             = fmt.Errorf("unable to stop instance, it must be running")
+	errUnableToOverrideConf             = fmt.Errorf("unable to override instance configuration, it must be setup")
+)
+
+type m3dbInst struct {
+	sync.Mutex
+	opts         Options
+	id           string
+	rack         string
+	zone         string
+	weight       uint32
+	endpoint     string
+	shards       shard.Shards
+	status       InstanceStatus
+	currentBuild build.ServiceBuild
+	currentConf  build.ServiceConfiguration
+	operator     operator.Operator
+	opListener   operator.Listener
+	opListenerID operator.ListenerID
+	m3dbClient   m3dbrpc.TChanNode
+}
+
+// NewM3DBInstance returns a new M3DBInstance.
+func NewM3DBInstance(
+	inst services.PlacementInstance,
+	op operator.Operator,
+	opts Options,
+) (M3DBInstance, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+	retInst := &m3dbInst{
+		opts:     opts,
+		id:       inst.ID(),
+		rack:     inst.Rack(),
+		zone:     inst.Zone(),
+		weight:   inst.Weight(),
+		endpoint: inst.Endpoint(),
+		shards:   inst.Shards(),
+		status:   InstanceStatusUninitialized,
+		operator: op,
+	}
+	if listener := opts.Listener(); listener != nil {
+		opListener := operator.NewListener(
+			func(s string) { listener.OnProcessTerminate(retInst, s) },
+			func(ts time.Time) { listener.OnHeartbeatTimeout(retInst, ts) },
+			func(s string) { listener.OnOverwrite(retInst, s) },
+		)
+		retInst.opListener = opListener
+		retInst.opListenerID = op.RegisterListener(opListener)
+	}
+	return retInst, nil
+}
+
+func (i *m3dbInst) String() string {
+	i.Lock()
+	defer i.Unlock()
+	return fmt.Sprintf(
+		"Instance[ID=%s, Rack=%s, Zone=%s, Weight=%d, Endpoint=%s, Shards=%s]",
+		i.id, i.rack, i.zone, i.weight, i.endpoint, i.shards.String(),
+	)
+}
+
+func (i *m3dbInst) ID() string {
+	i.Lock()
+	defer i.Unlock()
+	return i.id
+}
+
+func (i *m3dbInst) SetID(id string) services.PlacementInstance {
```

Comment:
Actually can't do this because `M3DBInstance` nests `services.PlacementInstance`, which requires the return type to be a `services.PlacementInstance`